packages/multidid/src/index.ts

```diff
@@ -0,0 +1,255 @@
+ * const multidid = Multidid.fromDIDString(didString)
+ *
+ * // Encode to bytes
+ * multidid.encode()
+ *
+ * // Decode from bytes to multidid instance
+ * Multidid.decode(bytes)
+ *
+ * // Encode as base16 string
+ * const mdStr = multidid.toString('base16')
+ * console.log(mdStr)
+ * // f9d1aed013b6a27bcceb6a42d62a3a8d02a6f0d73653215771de243a63ac048a18b59da29307a364d6b6954427a31796d75657041513448454859534631483871754735474c5656515233646a6458336d446f6f5770
+ *
+ * // Multidid instance from base encoded string
+ * Multidid.fromString(mdStr)
+ *
+ * // DID string from multidid
+ * multidid.toDidString()
+ *
+ * ```
+ *
+ * @module @didtools/multidid
+ */
+
+import * as u8a from 'uint8arrays'
+import { alloc } from 'uint8arrays/alloc'
+import { varint } from 'multiformats'
+import { base58btc } from 'multiformats/bases/base58'
+import { base16 } from 'multiformats/bases/base16'
+
+type SupportedBase = 'base16' | 'base58btc'
+
+const MULTIDID_CODEC = 3357
+const ALL_METHOD_CODE = 85
+const PKH_METHOD_CODE = 202
+
+const KEY_METHOD_CODES_LENGTH: Record<number, number> = {
+  // 0xe7
+  231: 33,
+  // 0xeb
+  235: 96,
+  //0xec
+  236: 32,
+  //0xed
+  237: 32,
+  //0x1200
+  4608: 33,
+  //0x1201
+  4609: 49,
+  //0x1202
+  4610: 67,
+  //0x1205
+  // TODO RSA key length
+  // 4613: 270 or 256
+}
+
+// TODO
+// Could makes sense to add general did codec interface
+// did:* encoding should be used explicitly? or fail when trying
+
+export class Multidid {
+  private mdBytes: Uint8Array | null
+
+  /**
+   * @param {number} code - DID Method Codec
+   * @param {Uint8Array} id - DID method id
+   * @param {Uint8Array} url - DID Method url portion
+   *
+   */
+  constructor(private code: number, private id: Uint8Array, private url: Uint8Array) {
+    this.mdBytes = null
+  }
+
+  /**
+   * Decoded a multidid from its binary representation
+   */
+  static decode(bytes: Uint8Array): Multidid {
+    const [, didCodeLen] = varint.decode(bytes, 0)
+    const [methodCode, methodCodeLen] = varint.decode(bytes, didCodeLen)
+    const methodIdOffset = didCodeLen + methodCodeLen
+
+    let methodIdLen
+    if (methodCode === ALL_METHOD_CODE) {
+      methodIdLen = 0
+    } else if (methodCode === PKH_METHOD_CODE) {
+      throw new Error('TODO')
+    } else {
+      methodIdLen = KEY_METHOD_CODES_LENGTH[methodCode]
+    }
+    if (!methodIdLen && methodIdLen !== 0) throw new Error('Not matching did method code found')
+
+    const methodId = alloc(methodIdLen)
+    methodId.set(bytes.slice(methodIdOffset, methodIdOffset + methodIdLen))
+    const urlLenOffset = methodIdOffset + methodIdLen
+    const [urlLen, urlLenLen] = varint.decode(bytes, urlLenOffset)
+    const urlBytesOffset = urlLenOffset + urlLenLen
+    const url = alloc(urlLen)
+    url.set(bytes.slice(urlBytesOffset, urlBytesOffset + urlLen))
+    return new Multidid(methodCode, methodId, url)
+  }
+
+  /**
+   * Encode multidid to bytes
+   */
+  encode(): Uint8Array {
```

Comment:
Can we make this method private? One accessor through `.bytes` should be enough.

Reply:
Do we really need `to_multibase` and `from_multibase` though?

Reply:
we use it for the tests
```
const didString = "did:example:123456"
const hexMD = 'f9d1a550e6578616d706c653a313233343536'

assert Multidid.from_multibase(hexMD).to_string() == didString
assert Multidid.from_string(didString).to_multibase('f') == hexMD
```
maby I just don't see the cost to having it

Reply:
Ok!